# Jessie Xie (21918545)

# Read data and Deal with categorical variables

> The purpose of the notebook is to 1) clean data, 2) deal with categorical data, 3) split train and test data.  <br>

In [1]:
# Import modules
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn import tree
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score
import graphviz 

In [2]:
# Read in the csv file
clean_absorbance = pd.read_csv('a_data.csv')
clean_absorbance.head()

Gender  AgeM   ECV  TPP    SC  TympPoF  OAE1  OAE1.4  OAE2  OAE2.8  ...  \
0       0   112  1.08    0  0.52        1   3.0    10.0  18.0    14.0  ...   
1       0   114  1.28    0  1.28        0   7.0     4.0   9.0     5.0  ...   
2       1   118  1.26    0  0.87        0   9.0    12.0   8.0    12.0  ...   
3       0   119  0.64    0  0.60        0  13.0     9.0  13.0    11.0  ...   
4       0    95  0.92    0  1.65        0   7.0     0.0   2.0    12.0  ...   

   f.6168.8433. f.6349.6042.  f.6535.6618. f.6727.1713.  f.6924.2925.  \
0        0.5370       0.4869        0.4301       0.4080        0.3804   
1        0.2636       0.2578        0.2443       0.2549        0.2604   
2        0.4269       0.4142        0.3910       0.3799        0.3740   
3        0.0077       0.0123        0.0066       0.0000        0.0126   
4        0.2332       0.2216        0.2228       0.2653        0.2596   

   f.7127.1897.  f.7336.0323.  f.7550.9945.  f.7772.2555.  f.8000.0000.  
0        0.3291        0.2951        0.2530        0.2282        0.2003  
1        0.2397        0.2299        0.2048        0.1917        0.1659  
2        0.3536        0.3473        0.3237        0.3056        0.2881  
3        0.0231        0.0405        0.0401        0.0391        0.0258  
4        0.2576        0.2838        0.3598        0.3953        0.3642  

[5 rows x 121 columns]

In [3]:
print(clean_absorbance.dtypes)
print(clean_absorbance.shape)

Gender            int64
AgeM              int64
ECV             float64
TPP               int64
SC              float64
                 ...   
f.7127.1897.    float64
f.7336.0323.    float64
f.7550.9945.    float64
f.7772.2555.    float64
f.8000.0000.    float64
Length: 121, dtype: object
(239, 121)


In [4]:
# View the statistic results of the dataset
clean_absorbance.describe()

Gender        AgeM         ECV         TPP          SC     TympPoF  \
count  239.000000  239.000000  239.000000  239.000000  239.000000  239.000000   
mean     0.435146  106.426778    0.998703    1.351464    0.716799    0.125523   
std      0.496817    8.507157    0.225526    5.153312    0.372878    0.332006   
min      0.000000   93.000000    0.530000    0.000000    0.000000    0.000000   
25%      0.000000  100.000000    0.845000    0.000000    0.500000    0.000000   
50%      0.000000  104.000000    0.960000    0.000000    0.625000    0.000000   
75%      1.000000  114.000000    1.130000    0.000000    0.845000    0.000000   
max      1.000000  129.000000    1.810000   51.000000    3.040000    1.000000   

             OAE1      OAE1.4        OAE2      OAE2.8  ...  f.6168.8433.  \
count  239.000000  239.000000  239.000000  239.000000  ...    239.000000   
mean    10.479498   13.055230   12.277824   10.297448  ...      0.236391   
std      6.409594    6.453945    5.661692    4.725506  ...      0.198270   
min      0.000000    0.000000    0.000000    0.000000  ...      0.000000   
25%      6.000000    9.000000    8.850000    6.950000  ...      0.073850   
50%      9.800000   13.000000   12.000000   10.700000  ...      0.207300   
75%     14.900000   17.550000   16.100000   14.000000  ...      0.362350   
max     31.300000   33.100000   27.500000   22.400000  ...      0.935300   

       f.6349.6042.  f.6535.6618.  f.6727.1713.  f.6924.2925.  f.7127.1897.  \
count    239.000000    239.000000    239.000000    239.000000    239.000000   
mean       0.215208      0.188059      0.185924      0.183527      0.168937   
std        0.186613      0.176799      0.174104      0.166406      0.159883   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.060250      0.029500      0.032900      0.039700      0.033550   
50%        0.193000      0.165800      0.154600      0.147700      0.133100   
75%        0.329050      0.288850      0.268550      0.276250      0.263550   
max        0.916400      0.897900      0.895400      0.879700      0.855800   

       f.7336.0323.  f.7550.9945.  f.7772.2555.  f.8000.0000.  
count    239.000000    239.000000    239.000000    239.000000  
mean       0.163462      0.159700      0.166354      0.155682  
std        0.154218      0.158385      0.162071      0.168058  
min        0.000000      0.000000      0.000000      0.000000  
25%        0.040650      0.026950      0.035650      0.015400  
50%        0.125500      0.118100      0.115500      0.109400  
75%        0.253050      0.253050      0.256650      0.249400  
max        0.816600      0.774000      0.772400      0.856400  

[8 rows x 119 columns]

In [5]:
# Check the null values
def check_null(df):
    df_null = pd.DataFrame({'num_null':df.isna().sum(), 'precent_null':df.isna().sum()/len(df)})
    return df_null[df_null["num_null"] > 0]

check_null(clean_absorbance)

Empty DataFrame
Columns: [num_null, precent_null]
Index: []

In [6]:
# Check how many numerical variables and categorical variables
columns = clean_absorbance.applymap(np.isreal).all()
print((columns).value_counts())

True     119
False      2
dtype: int64


> Comment: There are 119 numerical variables and 2 categorical variables

In [7]:
# Find the categorical variables
cols = clean_absorbance.columns
num_cols = clean_absorbance._get_numeric_data().columns
list(set(cols) - set(num_cols))

['OAEPoF', 'EarSide']

In [8]:
clean_absorbance[['OAEPoF', 'EarSide']]

OAEPoF EarSide
0        P    Left
1        P    Left
2        P    Left
3        P    Left
4        F    Left
..     ...     ...
234      P   Right
235      P   Right
236      P   Right
237      F   Right
238      P   Right

[239 rows x 2 columns]

In [9]:
# Encode the categorical variables
clean_absorbance['OAEPoF_encode'] = clean_absorbance['OAEPoF'].replace(['P', 'F'],[0,1]) # convert P to 0 and F to 1
clean_absorbance['EarSide_encode'] = clean_absorbance['EarSide'].replace(['Left', 'Right'],[0,1]) # convert Left to 0 and Right to 1
clean_absorbance[['OAEPoF_encode','OAEPoF','EarSide_encode', 'EarSide']]

OAEPoF_encode OAEPoF  EarSide_encode EarSide
0                0      P               0    Left
1                0      P               0    Left
2                0      P               0    Left
3                0      P               0    Left
4                1      F               0    Left
..             ...    ...             ...     ...
234              0      P               1   Right
235              0      P               1   Right
236              0      P               1   Right
237              1      F               1   Right
238              0      P               1   Right

[239 rows x 4 columns]

In [10]:
# Copy the data set 
df = clean_absorbance.copy()
df = df.drop(['OAEPoF','EarSide'], axis=1) # drop "OAEPoF" and "EarSide"
df

Gender  AgeM   ECV  TPP    SC  TympPoF  OAE1  OAE1.4  OAE2  OAE2.8  ...  \
0         0   112  1.08    0  0.52        1   3.0    10.0  18.0    14.0  ...   
1         0   114  1.28    0  1.28        0   7.0     4.0   9.0     5.0  ...   
2         1   118  1.26    0  0.87        0   9.0    12.0   8.0    12.0  ...   
3         0   119  0.64    0  0.60        0  13.0     9.0  13.0    11.0  ...   
4         0    95  0.92    0  1.65        0   7.0     0.0   2.0    12.0  ...   
..      ...   ...   ...  ...   ...      ...   ...     ...   ...     ...  ...   
234       1   104  1.18    0  1.07        0  10.3    18.6  12.0     0.5  ...   
235       0   104  0.78    0  0.77        0  13.2    23.5  17.3    14.3  ...   
236       1   106  0.90    2  1.02        0   5.3    11.0  15.3    14.6  ...   
237       1    99  0.76    0  0.15        1   0.0     0.0   0.0     3.0  ...   
238       1   118  0.66    0  1.18        1   8.3     9.9  16.6    15.7  ...   

     f.6535.6618.  f.6727.1713.  f.6924.2925.  f.7127.1897.  f.7336.0323.  \
0          0.4301        0.4080        0.3804        0.3291        0.2951   
1          0.2443        0.2549        0.2604        0.2397        0.2299   
2          0.3910        0.3799        0.3740        0.3536        0.3473   
3          0.0066        0.0000        0.0126        0.0231        0.0405   
4          0.2228        0.2653        0.2596        0.2576        0.2838   
..            ...           ...           ...           ...           ...   
234        0.1592        0.1549        0.1346        0.1250        0.1125   
235        0.4316        0.4108        0.3587        0.3216        0.2877   
236        0.4832        0.4501        0.3983        0.3545        0.3115   
237        0.2302        0.2029        0.1586        0.1279        0.0934   
238        0.5295        0.4979        0.4539        0.4229        0.3871   

     f.7550.9945.  f.7772.2555.  f.8000.0000.  OAEPoF_encode  EarSide_encode  
0          0.2530        0.2282        0.2003              0               0  
1          0.2048        0.1917        0.1659              0               0  
2          0.3237        0.3056        0.2881              0               0  
3          0.0401        0.0391        0.0258              0               0  
4          0.3598        0.3953        0.3642              1               0  
..            ...           ...           ...            ...             ...  
234        0.1242        0.1356        0.1372              0               1  
235        0.3026        0.2838        0.2682              0               1  
236        0.3136        0.2871        0.2679              0               1  
237        0.0612        0.0427        0.0031              1               1  
238        0.4081        0.3874        0.3852              0               1  

[239 rows x 121 columns]

In [11]:
X = df.drop(['OverallPoF'], axis=1)
print("X shape:",X.shape)
y = df[['OverallPoF']]
print("y shape:",y.shape)

X shape: (239, 120)
y shape: (239, 1)


In [12]:
# A 80/20 random split on the dataset to form a training set and a test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=12)
print("X train shape:",X_train.shape)
print("y train shape:",y_train.shape)
print("X test shape:",X_test.shape)
print("y test shape:",y_test.shape)

X train shape: (191, 120)
y train shape: (191, 1)
X test shape: (48, 120)
y test shape: (48, 1)


In [13]:
# save dataset
X_train.to_csv('X_train_J.csv', index=False)
y_train.to_csv('y_train_J.csv', index=False)
X_test.to_csv('X_test_J.csv', index=False)
y_test.to_csv('y_test_J.csv', index=False)
